src/pkg/engine/pepr/stream.go

```diff
@@ -0,0 +1,436 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023-Present The UDS Authors
+
+// Package pepr contains the logic monitoring Pepr operations in a UDS Cluster
+package pepr
+
+import (
+	"bufio"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/defenseunicorns/uds-cli/src/pkg/style"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// StreamKind represents the type of Pepr stream
+type StreamKind string
+
+type StreamReader struct {
+	JSON            bool
+	FilterStream    StreamKind
+	filterNamespace string
+	filterName      string
+	showTimestamp   bool
+	indent          string
+	lastEntryHeader string
+	lastEntryBody   string
+	repeatCount     int
+	mutex           sync.Mutex
+}
+
+// LogEntry represents a log entry from Pepr
+type LogEntry struct {
+	Level     int    `json:"level"`
+	Time      int64  `json:"time"`
+	Pid       int    `json:"pid"`
+	Hostname  string `json:"hostname"`
+	Uid       string `json:"uid"`
+	Namespace string `json:"namespace"`
+	Name      string `json:"name"`
+	Res       struct {
+		Uid     string `json:"uid"`
+		Allowed bool   `json:"allowed"`
+		Status  struct {
+			Code    int    `json:"code"`
+			Message string `json:"message"`
+		} `json:"status"`
+		// Patch and PatchType are only present for mutations
+		Patch     *string `json:"patch,omitempty"`
+		PatchType *string `json:"patchType,omitempty"`
+	} `json:"res"`
+	Msg string `json:"msg"`
+	// Metadata is only present for operator logs
+	Metadata *struct {
+		Name      string `json:"name"`
+		Namespace string `json:"namespace"`
+	}
+}
+
+// operation represents a JSON Patch operation
+type operation struct {
+	Op    string          `json:"op"`
+	Path  string          `json:"path"`
+	Value json.RawMessage `json:"value"`
+}
+
+const (
+	// AnyStream represents all Pepr logs
+	AnyStream StreamKind = ""
+	// PolicyStream represents all Pepr admission controller logs
+	PolicyStream StreamKind = "policies"
+	// OperatorStream represents all UDS Operator logs
+	OperatorStream StreamKind = "operator"
+	// AllowStream represents all Pepr admission controller allow logs
+	AllowStream StreamKind = "allowed"
+	// DenyStream represents all Pepr admission controller deny logs
+	DenyStream StreamKind = "denied"
+	// MutateStream represents all Pepr admission controller mutation logs
+	MutateStream StreamKind = "mutated"
+	// FailureStream represents all admission controller deny logs and operator failure logs
+	FailureStream StreamKind = "failed"
+)
+
+// NewStreamReader creates a new PeprStreamReader
+func NewStreamReader(timestamp bool, filterNamespace, filterName string) *StreamReader {
+	// Use a longer indent when timestamps are enabled
+	indent := ""
+	if timestamp {
+		indent = "                     "
```

Comment:
~This is fine for now and we can make a follow up issue, but it'd be nice to get the size of the terminal and make this a bit more responsive in smaller terminals~

I take it back, this looks fine even in small terminals (also they can just resize and it'll be fine)

Reply:
@Madeline-UX and I discussed that a lot, first thing is they don't have to enable timestamps and it's more narrow, second it does wrap and we try to add enough line breaks to make it acceptable still.